File: ruby/ruby_on_rails/spec/controller/tasks_controller_spec.rb
RSpec.describe TasksController, type: :request do
  let(:title) { "My task" }
  let(:user) { create(:user) }

  describe "GET /tasks" do
    context "when a User is authenticated" do
      before do
        sign_in user
        create(:task, user:)
        get "/tasks"
      end

      it "returns a JSON array" do
        expect(response.body).to be_a String
        expect(response.parsed_body).to all have_key "id"
      end

      it "returns a list of Task" do
        data = response.parsed_body

        data.each do |task|
          %w[id title].each do |key|
            expect(task).to have_key key
          end
        end
      end

      it "returns a ok HTTP status" do
        expect(response).to have_http_status :ok
      end
    end

    context "when a User is not authenticated" do
      it "returns a unauthorized HTTP status" do
        get "/tasks"

        expect(response).to have_http_status :unauthorized
      end
    end
  end

  describe "POST /tasks" do
    context "when a User is authenticated" do
      before do
        sign_in user
      end

      context "with proper params" do
        before do
          post "/tasks", params: { task: { title: } }
        end

        it "returns a JSON object" do
          expect(response.body).to be_a String
          expect(response.parsed_body).to have_key "id"
        end

        it "creates an instance of Task" do
          expect(Task.count).to eq 1
        end

        it "returns the new instance of Task" do
          data = response.parsed_body
          expect(data["title"]).to eq title
        end

        it "returns a created HTTP status" do
          expect(response).to have_http_status :created
        end
      end

      context "without proper params" do
        before do
          post "/tasks", params: { task: { title: nil } }
        end

        it "returns a JSON object" do
          expect(response.body).to be_a String
          expect(response.parsed_body).to have_key "errors"
        end

        it "does not create an instance of Task" do
          expect(Task.count).to eq 0
        end

        it "returns a list of error messages" do
          data = response.parsed_body
          expect(data["errors"]).to match({ title: "Title can't be blank" })
        end

        it "returns a unprocessable_entity HTTP status" do
          expect(response).to have_http_status :unprocessable_entity
        end
      end
    end

    context "when a User is not authenticated" do
      it "returns a unauthorized HTTP status" do
        post "/tasks", params: { task: { title: } }

        expect(response).to have_http_status :unauthorized
      end
    end
  end

  describe "DELETE /tasks/:id" do
    let(:task) { create(:task, user:) }

    context "when a User is authenticated" do
      before do
        sign_in user
      end

      context "when the Task is found" do
        before do
          delete "/tasks/#{task.id}"
        end

        it "destroys the instance of Task" do
          expect(Task.count).to be 0
        end

        it "returns a ok HTTP status" do
          expect(response).to have_http_status :ok
        end
      end

      context "when the Task is not found" do
        before do
          delete "/tasks/#{task.id + 1}"
        end

        it "returns a JSON object" do
          expect(response.body).to be_a String
          expect(response.parsed_body).to have_key "errors"
        end

        it "returns a list of error messages" do
          data = response.parsed_body
          expect(data["errors"]).to match({ task: "Task must exist" })
        end

        it "returns a not_found HTTP status" do
          expect(response).to have_http_status :not_found
        end
      end
    end

    context "when a User is not authenticated" do
      it "returns a unauthorized HTTP status" do
        delete "/tasks/#{task.id}"

        expect(response).to have_http_status :unauthorized
      end
    end
  end

  describe "PATCH /tasks/:id/complete" do
    let(:task) { create(:task, user:) }

    context "when a User is authenticated" do
      before do
        sign_in user
      end

      context "when the Task is found" do
        before do
          patch "/tasks/#{task.id}/complete"
        end

        it "returns a JSON object" do
          expect(response.body).to be_a String
          expect(response.parsed_body).to have_key "id"
        end

        it "returns the instance of Task" do
          data = response.parsed_body
          expect(data["id"]).to be task.id
        end

        it "marks the Task as completed" do
          expect(Task.find(task.id)).to be_completed
        end

        it "returns a ok HTTP status" do
          expect(response).to have_http_status :ok
        end
      end

      context "when the Task is not found" do
        before do
          patch "/tasks/#{task.id + 1}/complete"
        end

        it "returns a JSON object" do
          expect(response.body).to be_a String
          expect(response.parsed_body).to have_key "errors"
        end

        it "returns a list of error messages" do
          data = response.parsed_body
          expect(data["errors"]).to match({ task: "Task must exist" })
        end

        it "returns a not_found HTTP status" do
          expect(response).to have_http_status :not_found
        end
      end
    end

    context "when a User is not authenticated" do
      it "returns a unauthorized HTTP status" do
        patch "/tasks/#{task.id}/complete"

        expect(response).to have_http_status :unauthorized
      end
    end
  end
end

File: node/express/spec/controllers/tasks.controller.spec.ts
import type { Prisma } from "@prisma/client"
import type { User } from "src/models/user.model.ts"
import supertest from "supertest"
import { beforeEach } from "vitest"
import app from "src/app.ts"
import prismaTask from "src/models/task.model.ts"
import prismaUser from "src/models/user.model.ts"

describe("TasksController", () => {
  let response: supertest.Response
  const request = supertest(app)
  const title = "My task"
  let user: User

  const createTask = async () => await prismaTask.task.create({ data: { title, userId: user.id } })
  const createUser = async () => await prismaUser.user.create({
    data: { email: "user@example.com", password: "password" }
  })

  describe("GET /tasks", () => {
    beforeEach(async () => {
      user = await createUser()
      await createTask()
      response = await request.get("/tasks")
    })

    describe("when a User is authenticated", () => {
      beforeEach(async () => {
        response = await request.get("/tasks").set("Authorization", user.jwt)
      })

      it("returns a JSON type response", () => {
        expect(response.type).toBe("application/json")
      })

      it("returns a list of Task", () => {
        expect(response.body).toMatchObject([{ id: 1, title }])
      })

      it("returns a ok HTTP status", () => {
        expect(response.status).toBe(200)
      })
    })

    describe("when a User is not authenticated", () => {
      beforeEach(async () => {
        response = await request.get("/tasks")
      })

      it("returns a unauthorized HTTP status", () => {
        expect(response.status).toBe(401)
      })
    })
  })

  describe("POST /tasks", () => {
    describe("when a User is authenticated", () => {
      beforeEach(async () => {
        user = await createUser()
      })

      describe("with proper params", () => {
        beforeEach(async () => {
          response = await request.post("/tasks").set("Authorization", user.jwt).send({ task: { title } })
        })

        it("return a JSON type response", () => {
          expect(response.type).toBe("application/json")
        })

        it("creates an instance of Task", async () => {
          await expect(prismaTask.task.count()).resolves.toBe(1)
        })

        it("returns the new instance of Task", () => {
          expect(response.body).toMatchObject({ title })
        })

        it("returns a created HTTP status", () => {
          expect(response.status).toBe(201)
        })
      })

      describe("without proper params", () => {
        beforeEach(async () => {
          response = await request.post("/tasks").set("Authorization", user.jwt).send({ task: { title: null } })
        })

        it("return a JSON type response", () => {
          expect(response.type).toBe("application/json")
        })

        it("does not create an instance of Task", async () => {
          await expect(prismaTask.task.count()).resolves.toBe(0)
        })

        it("returns a list error messages", () => {
          expect(response.body).toStrictEqual({ errors: { title: "Title can't be blank" } })
        })

        it("returns a unprocessable_entity HTTP status", () => {
          expect(response.status).toBe(422)
        })
      })
    })

    describe("when a User is not authenticated", () => {
      beforeEach(async () => {
        response = await request.post("/tasks").send({ task: { title } })
      })

      it("returns a unauthorized HTTP status", () => {
        expect(response.status).toBe(401)
      })
    })
  })

  describe("DELETE /tasks/:id", () => {
    let task: Prisma.TaskGetPayload<true>

    beforeEach(async () => {
      user = await createUser()
      task = await createTask()
    })

    describe("when a User is authenticated", () => {
      describe("when the Task is found", () => {
        beforeEach(async () => {
          response = await request.delete(`/tasks/${task.id}`).set("Authorization", user.jwt)
        })

        it("destroys the instance of Task", async () => {
          await expect(prismaTask.task.count()).resolves.toBe(0)
        })

        it("returns a ok http status", () => {
          expect(response.status).toBe(200)
        })
      })

      describe("when the Task is not found", () => {
        beforeEach(async () => {
          response = await request.delete(`/tasks/${task.id + 1}`).set("Authorization", user.jwt)
        })

        it("return a JSON type response", () => {
          expect(response.type).toBe("application/json")
        })

        it("returns a list error messages", () => {
          expect(response.body).toStrictEqual({ errors: { task: "Task must exist" } })
        })

        it("returns a not_found HTTP status", () => {
          expect(response.status).toBe(404)
        })
      })
    })

    describe("when a User is not authenticated", () => {
      beforeEach(async () => {
        response = await request.delete(`/tasks/${task.id}`)
      })

      it("returns a unauthorized HTTP status", () => {
        expect(response.status).toBe(401)
      })
    })
  })

  describe("PATCH /tasks/:id/complete", () => {
    let task: Prisma.TaskGetPayload<true>

    beforeEach(async () => {
      user = await createUser()
      task = await createTask()
    })

    describe("when a User is authenticated", () => {
      describe("when the Task is found", () => {
        beforeEach(async () => {
          response = await request.patch(`/tasks/${task.id}/complete`).set("Authorization", user.jwt)
        })

        it("returns a JSON type response", () => {
          expect(response.type).toBe("application/json")
        })

        it("returns the instance of Task", () => {
          expect(response.body).toMatchObject({ id: task.id })
        })

        it("marks the Task as completed", async () => {
          await expect(prismaTask.task.findUnique({ where: { id: task.id } }))
            .resolves
            .toMatchObject({ completed: true })
        })

        it("returns a ok http status", () => {
          expect(response.status).toBe(200)
        })
      })

      describe("when the Task is not found", () => {
        beforeEach(async () => {
          response = await request.delete(`/tasks/${task.id + 1}`).set("Authorization", user.jwt)
        })

        it("return a JSON type response", () => {
          expect(response.type).toBe("application/json")
        })

        it("returns a list error messages", () => {
          expect(response.body).toStrictEqual({ errors: { task: "Task must exist" } })
        })

        it("returns a not_found HTTP status", () => {
          expect(response.status).toBe(404)
        })
      })
    })

    describe("when a User is not authenticated", () => {
      beforeEach(async () => {
        response = await request.patch(`/tasks/${task.id}/complete`)
      })

      it("returns a unauthorized HTTP status", () => {
        expect(response.status).toBe(401)
      })
    })
  })
})
